posts/2021/11/pydata-extensibility-vision.md

```diff
@@ -0,0 +1,363 @@
+<!--
+.. title: A vision for extensibility to GPU & distributed support for SciPy, scikit-learn, scikit-image and beyond
+.. slug: pydata-extensibility-vision
+.. date: 2021-11-15 10:00:00 UTC-07:00
+.. author: Ivan Yashchuk, Ralf Gommers
+.. tags: SciPy, NumPy, CuPy, scikit-learn, scikit-image, uarray, Array API
+.. category:
+.. link:
+.. description:
+.. type: text
+.. previewimage: /images/2021/11/nep-0047-library-dependencies.png
+-->
+
+
+Quansight Labs is partnering with AMD to develop a new extensibility system for
+the PyData stack. This work will enable the PyData libraries to be executed on a
+GPU, as well as running in distributed mode across multiple GPUs. In this blog
+post, we talk about the motivation behind the proposal and describe our vision
+for making array libraries more interoperable.
+
+With the proposed dispatch layers, the downstream libraries would be enabled to
+use multiple kinds of arrays, without having to have a hard dependency on all of
+those array libraries. We want to have a unified design using the same tools
+applicable across SciPy, scikit-learn, scikit-image and other libraries. In this
+situation, it's best to have a discussion spanning multiple projects.
+<!-- This post is the invitation to discuss the idea on this forum https://discuss.scientific-python.org/ -->
+
+
+<p align="center">
+    <img
+     alt="A diagram showing possible interconnections that would be enabled by Array API."
+     src="/images/2021/11/nep-0047-library-dependencies.png">
+    <i><br>Possible interconnections that Array API enables.</i>
+</p>
+
+<!-- TEASER_END -->
+
+Over the years array computing in Python has evolved to support distributed
+arrays, GPU arrays, and other various kinds of arrays that work with specialized
+hardware, or carry additional metadata, or use different internal memory
+representations. The foundational library for array computing in the PyData
+ecosystem is NumPy. But NumPy alone is a CPU-only library - and a
+single-threaded one at that - and in a world where it's possible to get a GPU or
```

Comment:
From my understanding, NumPy with OpenBLAS can take advantage of multiple threads?

Reply:
I think I would remove "singled-threaded one at that", so we do not need to go through the details of "when numpy uses BLAS".

Reply:
Only the `linalg` functions that call into OpenBLAS or MKL. Which is only a few handfuls out of O(1000) functions/methods. I think it's fine to say this, NumPy really is single-threaded. It's good at releasing the GIL, but that's about it.

Reply:
Yes, that's what I meant, OpenBLAS/MKL is not a part of NumPy, it's a dependency that can be multithreaded.

Reply:
Let's leave this as is, should be okay.